daemon/src/taker_cfd.rs

```diff
@@ -198,61 +187,78 @@ where
             .await?;
         Ok(())
     }
-}
 
-#[xtra_productivity]
-impl<O, M, W> Actor<O, M, W> {
-    async fn handle_propose_settlement(&mut self, msg: ProposeSettlement) -> Result<()> {
+    async fn handle_propose_settlement(
+        &mut self,
+        msg: ProposeSettlement,
+        ctx: &mut xtra::Context<Self>,
+    ) -> Result<()> {
         let ProposeSettlement {
             order_id,
             current_price,
         } = msg;
 
+        let disconnected = self
+            .collab_settlement_actors
+            .get_disconnected(order_id)
+            .with_context(|| format!("Settlement for order {} is already in progress", order_id))?;
+
         let mut conn = self.db.acquire().await?;
         let cfd = load_cfd_by_order_id(order_id, &mut conn).await?;
 
-        if !cfd.is_collaborative_settle_possible() {
-            anyhow::bail!(
-                "Settlement proposal not possible because for cfd {} is in state {} which cannot be collaboratively settled",
-                order_id,
-                cfd.state
-            )
-        }
+        let this = ctx
+            .address()
+            .expect("actor to be able to give address to itself");
+        let (addr, fut) = collab_settlement_taker::Actor::new(
+            cfd,
+            self.projection_actor.clone(),
+            this,
+            current_price,
+            self.conn_actor.clone(),
+            self.n_payouts,
+        )?
+        .create(None)
+        .run();
 
-        let proposal = cfd.calculate_settlement(current_price, self.n_payouts)?;
+        disconnected.insert(addr);
+        self.tasks.add(fut);
 
-        if self
-            .current_pending_proposals
-            .contains_key(&proposal.order_id)
-        {
-            anyhow::bail!(
-                "Settlement proposal for order id {} already present",
-                order_id
-            )
-        }
+        Ok(())
+    }
 
-        let new_proposal = UpdateCfdProposal::Settlement {
-            proposal: proposal.clone(),
-            direction: SettlementKind::Outgoing,
+    async fn handle_settlement_completed(
+        &mut self,
+        msg: collab_settlement_taker::Completed,
+    ) -> Result<()> {
+        let (order_id, settlement) = match msg {
+            collab_settlement_taker::Completed::Confirmed {
+                order_id,
+                settlement,
+            } => (order_id, settlement),
+            collab_settlement_taker::Completed::Rejected { .. } => {
+                return Ok(());
+            }
+            collab_settlement_taker::Completed::Failed { order_id, error } => {
+                tracing::warn!(%order_id, "Collaborative settlement failed: {:#}", error);
```

Comment:
:question: I ask this for myself mainly: why did you choose `warn` over `error` here?

Reply:
Should probably be an `error`? :D

Reply:
Actually: https://github.com/itchysats/itchysats/pull/743#discussion_r759831579.